Added EMS On/Off Supervisory control for air loop water coils

rraustad opened the pull request:
## Pull request overview

EMS On/Off Supervisory control is intended to actuate plant loop components including water coils. These air loop components were not being controlled correctly.  Addresses #5650 
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [ ] Code style (parentheses padding, variable names)
- [ ] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [ ] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

rraustad committed:
Added EMS On/Off Supervisory control for air loop water coils

rraustad commented on src/EnergyPlus/PlantUtilities.cc:
```diff
@@ -578,6 +578,13 @@ namespace PlantUtilities {
+						Node( NodeNum ).MassFlowRateRequest = 0.0;
+					}
+				}
+
```
@EnergyArchmage this does work for water coils. Anything look suspect?

I'll change this to:

```
if ( comp.EMSLoadOverrideOn && comp.EMSLoadOverrideValue == 0.0 ) {  
    Node( NodeNum ).MassFlowRate = 0.0;  
    Node( NodeNum ).MassFlowRateRequest = 0.0;
} else {
   Node( NodeNum ).MassFlowRate = a_node_MasFlowRate;  
   Node( NodeNum ).MassFlowRateRequest = a_node_MasFlowRateRequest; 
} 
```

EnergyArchmage replied:
This does look suspect to me.  It will be okay if there is only one device on the branch (which is usually the case for coils). This routine is intended to make sure all the components on a single branch get the same flow and request as the actuated node.  This code change could mess that up with one component getting different rates than others on the same branch.  Would have to scan all the components on the branch and see if any one is shut down, and then set the same values on everything on the branch.

rraustad commented:
Review of corrected control.

![ems_onoffsupervisory](https://cloud.githubusercontent.com/assets/5778156/15121108/eec0b0a2-15e6-11e6-9c96-69baf137576c.png)

rraustad committed:
When using component EMS On/OffSupervisory control account for more than 1 component on the branch

rraustad commented:
@EnergyArchmage @Myoldmopar This should be ready now.

EnergyArchmage commented on src/EnergyPlus/PlantUtilities.cc:
```diff
@@ -573,21 +574,29 @@ namespace PlantUtilities {
+			// check to see if any component on branch uses EMS On/Off Supervisory control to shut down flow
+			for ( int CompNum = 1, CompNum_end = branch.TotalComponents; CompNum <= CompNum_end; ++CompNum ) {
+				auto const & comp( branch.Comp( CompNum ) );
+				if ( comp.EMSLoadOverrideOn ) EMSLoadOverride = true;
```
Doesn't this need to check the value as well?   ( comp.EMSLoadOverrideValue == 0.0 )

rraustad replied:
Yes it does, not sure how I left that out.

rraustad commented:
I did 2 things, corrected the IF test to include the value, and also moved the check for EMS up into the FlowLock = Unlocked so as not to mess with the plant when the flow is locked.

I tested this EMS program and it worked as intended.

```
EnergyManagementSystem:Program,
  Heating_dispatch_Values, !- Name
  if (CurrentTime > 10.0),
    SET SPACE1_1_HeatingCoil_Enable = 0,
  ELSE,
    SET SPACE1_1_HeatingCoil_Enable = null,
  ENDIF,
  if (CurrentTime > 16.0),
    SET Main_CoolingCoil_Enable = 0,
  ELSE,
    SET Main_CoolingCoil_Enable = null,
  ENDIF;
```

rraustad committed:
Final code revisions and minor doc changes

EnergyArchmage commented:
okay, I have reviewed the code and it seems okay.  Snce the issue here was filed for coils it seems okay to just keep the changes in SetActuatedBranchFlowRate, but if we really want these actuators to work everywhere I suppose these kind of changes could also go into the more heavily used SetComponentFlowRate.    I am not sure.

rraustad committed:
Added EMS component control in SetComponentFlowRate per BG comment

rraustad commented:
I tested a fan coil example file, as @EnergyArchmage eluded, EMS control did not work for that component since it called SetComponentFlowRate to set water flow rate (why 2 functions?). In fact many equipment models use this function instead. Added logic to this other routine to expand reach of EMS plant component control.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5650-Airloop-water-coils-do-not-respond-to-EMS-On/Off-Supervisory-control

EnergyArchmage commented:
@rraustad asks "(why 2 functions?)" If I recall correctly, the water coil controller doesn't know what the outlet node is for its coils, so it can't pass the outlet node number to SetComponentFlowRate.  The old original idea was that the coil controller only needed to actuate the coil water inlet node.  But because one plant inlet node can be another device's outlet node, plant now looks at the request at the inlet but needs to set flow info on the device's outlet node as well.   SetActuatedBranchFlowRate deals with this by setting flow info on all the nodes on the branch to match what the controller is asking for.

EnergyArchmage commented on tst/EnergyPlus/unit/EMSManager.unit.cc:
```diff
@@ -323,6 +323,151 @@ TEST_F( EnergyPlusFixture, SupervisoryControl_Component ) {
+		EXPECT_EQ(Node( 3 ).MassFlowRateRequest, NodeMdot );
+
+}
+
```
I really like this unit test, really cool blend of idf input and building out a simple plant loop data structure.  

The only comment i have is that I think now we want all new unit tests to have a brief set of comments with some sort of narrative describing what the test is trying to accomplish.

rraustad commented:
I think these diff's are because this branch is behind develop. I expect no diff's unless this specific EMS actuator is used, which I think none of the example files use. I might use this as an excuse to comment the unit test as suggested.

rraustad committed:
Merge branch 'develop' of https://github.com/NREL/EnergyPlus into #5650-Airloop-water-coils-do-not-respond-to-EMS-On/Off-Supervisory-control

rraustad committed:
Add comments to unit test

rraustad commented:
Not sure what to say about the example file with large diff's. Certainly the answer is different, however, at the same time much the same. There are no diff's in the eio, particularly with warmup convergence. The change in answer, in my opinion, has nothing to do with the fix for this issue.

![pipeheattransfer_issue](https://cloud.githubusercontent.com/assets/5778156/15366794/01a2d952-1cf4-11e6-87b0-419d30a6ac54.png)

rraustad commented:
@Myoldmopar @EnergyArchmage this should be good to go.

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into #5650-Airloop-water-coils-do-not-respond-to-EMS-On/Off-Supervisory-control

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into #5650-Airloop-water-coils-do-not-respond-to-EMS-On/Off-Supervisory-control

EnergyArchmage committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into #5650-Airloop-water-coils-do-not-respond-to-EMS-On/Off-Supervisory-control

EnergyArchmage commented:
Okay I have reviewed the code changes again.  I reinvestigated the 5 files that are showing diffs.  I cannot explain them but I agree that the files run fine.  I ran local regressions and found them to be reproducible but also found no diffs in the annual results.  going to merge this in.
